src/parser.js

```diff
@@ -598,27 +598,55 @@ define([
     }
 
     function parseControlTree (form, controlsTree) {
-        function eachFunc(cEl, parentMug) {
+        function eachFunc(cEl, parentMug, index) {
             var $cEl = $(cEl),
-                mug = parseControlElement(form, $cEl, parentMug);
+                mug = parseControlElement(form, $cEl, parentMug),
+                tree = form.tree,
+                node = tree.getNodeFromMug(mug);
+
+            if (!node) {
+                mug.options.isControlOnly = true;
+                form.tree.insertMug(mug, 'into', parentMug);
+                node = tree.getNodeFromMug(mug);
+            }
 
-            form.controlTree.insertMug(mug, 'into', parentMug);
             if (mug.options.isControlOnly) {
                 // HACK fix abstraction broken by direct tree insert
                 form.mugMap[mug.ufid] = mug;
             }
 
-            if (mug.__className === "ReadOnly") {
-                return;
+            // preserve the control order 
+            var siblings = node.parent.getChildren(),
+                realIndex = index,
+                numControlStructs = 0,
+                numChildren = siblings.length;
+
+            for (var i = 0; i < numChildren; i++) {
+                if (siblings[i].value.options.isDataOnly) {
+                    realIndex++;
+                } else {
+                    numControlStructs++;
+                }
+
+                if (numControlStructs > index) {
+                    break;
+                }
             }
-            if (mug.options.controlNodeChildren) {
+
+            form.tree.insertMug(mug, 'index', parentMug, realIndex);
+
+            if (mug.__className !== "ReadOnly" && mug.options.controlNodeChildren) {
+                var index1 = 0;
```

Comment:
Very minor: might want to rename this to `child_index`. I would not even bring this up except it made code review a little harder. It wasn't easy to spot the difference between `index` and `index1`.